reconcile/ocm_update_recommended_version.py

```diff
@@ -0,0 +1,123 @@
+import functools
+from typing import Any
+
+import semver
+from reconcile import queries, mr_client_gateway
+from reconcile.ocm.types import OCMSpec
+from reconcile.utils.mr.ocm_update_recommended_version import (
+    CreateOCMUpdateRecommendedVersion,
+    UpdateInfo,
+    WorkloadRecommendedVersion,
+)
+from reconcile.utils.ocm import OCMMap
+
+QONTRACT_INTEGRATION = "ocm-update-recommended-version"
+
+
+def get_highest(version_set: set[str]) -> str:
+    sorted_version_set = sorted(
+        version_set, key=functools.cmp_to_key(semver.compare), reverse=True
+    )
+    return sorted_version_set[0]
+
+
+def get_majority(version_set: set[str], versions: list[str]) -> str:
+    return max(version_set, key=versions.count)
+
+
+def recommended_version(
+    versions: list[str], high_weight: int = 0, majority_weight: int = 0
+) -> str:
+    version_set = set(versions)
+    version_count = {k: versions.count(k) for k in version_set}
+
+    highest = get_highest(version_set)
+    majority = get_majority(version_set, versions)
+
+    high_value = version_count[highest] * high_weight
+    major_value = version_count[majority] * majority_weight
+
+    if major_value > high_value:
+        return majority
+
+    return highest
+
+
+def get_version_weights(ocm: dict[str, Any]) -> tuple[int, int]:
+    rv_weight = ocm.get("recommendedVersionWeight")
+    high_weight = 1
+    majority_weight = 1
+    if rv_weight:
+        if rv_weight.get("highest") is not None:
+            high_weight = rv_weight["highest"]
+        if rv_weight.get("majority") is not None:
+            majority_weight = rv_weight["majority"]
+    return high_weight, majority_weight
+
+
+def get_updated_recommended_versions(
+    ocm_info: dict[str, Any], cluster: dict[str, OCMSpec]
+) -> list[dict[str, str]]:
+
+    rv_current = ocm_info.get("recommendedVersions") or []
+
+    if len(rv_current) == 0:
+        return []
+
+    high_weight, majority_weight = get_version_weights(ocm_info)
+
+    rv_updated: list[dict[str, str]] = []
+    for workload in ocm_info["upgradePolicyAllowedWorkloads"] or []:
+        rv_workload = [rv for rv in rv_current if rv["workload"] == workload]
+        if len(rv_workload) == 1:
+            cluster_workload = [
+                c["name"]
+                for c in ocm_info["upgradePolicyClusters"]
+                if workload in c["upgradePolicy"]["workloads"]
+            ]
+            versions = [
+                cluster[k].spec.version for k in cluster if k in cluster_workload
+            ]
+            rv_update = rv_workload[0]
+            if len(versions) > 0:
+                rv_update["recommendedVersion"] = recommended_version(
+                    versions, high_weight, majority_weight
+                )
+            rv_updated.append(rv_update)
+        elif len(rv_workload) > 1:
+            raise ValueError("Expecting one recommended Version per workload!")
```

Comment:
what if `en(rv_workload) == 0`? Meaning we did not recommend a version yet for that workload.
If there is now at least one cluster running this workload, we should be able to recommend a version now, right ?

Reply:
Right, my idea was, one would have to add a workload to the recommendedVersion list first, only the it would get updated

Reply:
This does not seem to match with the dynamic nature of the fleet IMHO. Can we take this scenario into consideration ?

Reply:
Only when `recommendedVersionWeight` is set ?

Reply:
No, weight is optional. If not set it will be 1 for both.

Reply:
hmm i thought that default was fine initially, but it indeed presumes all orgs would have some recommended versions.  Should the default be 0 instead? and if there are no positive score, no version recommended ?